jina/clients/__init__.py

```diff
@@ -18,16 +31,130 @@ def py_client(**kwargs) -> 'PyClient':
         from jina.clients import py_client
 
         # to test connectivity
-        py_client(host='192.168.1.100', port_expose=55555).dry_run()
+        await py_client(host='192.168.1.100', port_expose=55555).dry_run()
 
         # to search
-        py_client(host='192.168.1.100', port_expose=55555).search(input_fn, output_fn)
+        await py_client(host='192.168.1.100', port_expose=55555).search(input_fn, output_fn)
 
         # to index
-        py_client(host='192.168.1.100', port_expose=55555).index(input_fn, output_fn)
+        await py_client(host='192.168.1.100', port_expose=55555).index(input_fn, output_fn)
+
+    .. note::
+        to perform `index`, `search` or `train`, py_client needs to be awaited, as it is a coroutine
+
     """
     from ..parser import set_client_cli_parser
     from ..helper import get_parsed_args
     from .python import PyClient
     _, args, _ = get_parsed_args(kwargs, set_client_cli_parser(), 'Client')
     return PyClient(args)
+
+
+def py_client_runtime(mode, input_fn, output_fn, **kwargs) -> None:
+    """ This method allows callback execution of client process in main process
+
+    PyClient writes the response from servicer in a PAIR-BIND socket.
+    Main process reads from it using PAIR-CONNECT.
+
+    """
+    from zmq.error import Again
+    from ..parser import set_client_cli_parser
+    from ..helper import get_parsed_args
+    _, args, _ = get_parsed_args(kwargs, set_client_cli_parser(), 'Client')
+
+    # setting this manually to avoid exposing on CLI
+    args.runtime = 'process'
+
+    with JinaLogger(context='PyClientRuntime') as logger:
+        with CtrlZmqlet(args=args, logger=logger, is_bind=False, is_async=False, timeout=10) as zmqlet:
+            with PyClientRuntime(args, mode=mode, input_fn=input_fn, output_fn=output_fn,
+                                 address=zmqlet.address, **kwargs):
+                counter = 0
+                while True:
+                    try:
+                        msg = zmqlet.sock.recv()
+                        if msg == b'TERMINATE':
```

Comment:
when is the client sending `TERMINATE` I haven't seen this happen, it is done between Peas but not from client right?

Reply:
`PyClient` would send a terminate signal to `PyClientRuntime`

Reply:
it should be the other way around right? PyClient does not know it lives inside a `PyClientRuntime`? Why don't u use `multiprocessing Events`?

Reply:
`PyClient` receives a stream of responses from gRPC servicer. To make sure the `callback` functions are executed back in the main process, `PyClient` sends each such response to `PyClientRuntime`. Once the stream is completed, `PyClient` sends a `TERMINATE` signal to `PyClientRuntime`. But you're right, this can be handled using an `Event.